streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -121,84 +121,88 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
-        case State.Error(e)                    => ZIO.halt(e)
-        case State.End                         => UIO.succeed((UIO.succeed(true), State.End))
+          case State.BatchMiddle(state, notifyProducer) =>
+            // The logic here is the same as the Empty state, except we don't need
+            // to notify the consumer on the transition
+            for {
+              step <- sink.step(state, a)
+              result <- if (sink.cont(step))
+                         UIO.succeed((UIO.succeed(true), State.BatchMiddle(step, notifyProducer)))
+                       else
+                         UIO.succeed((notifyProducer.await.as(true), State.BatchEnd(step, notifyProducer)))
+            } yield result
+
+          // The producer shouldn't actually see these states, but we still use sane
+          // transitions here anyway.
+          case s @ State.BatchEnd(_, batchTaken) => UIO.succeed((batchTaken.await.as(true), s))
+          case State.Error(e)                    => ZIO.halt(e)
+          case State.End                         => UIO.succeed((UIO.succeed(true), State.End))
+        })
       }.flatten
 
     // This function is used in an unfold, so `None` means stop consuming
     def consume(stateVar: Ref[State], permits: Semaphore): ZIO[R1, Option[E1], Chunk[B]] =
-      withStateVar(stateVar, permits) {
-        // If the state is empty, wait for a notification from the producer
-        case s @ State.Empty(_, notify) => UIO.succeed((notify.await.as(Chunk.empty), s))
-
-        case s @ State.Leftovers(_, _, notify) => UIO.succeed((notify.await.as(Chunk.empty), s))
-
-        case State.BatchMiddle(state, notifyProducer) =>
-          (for {
-            initial        <- sink.initial
-            notifyConsumer <- Promise.make[Nothing, Unit]
-            extractResult  <- sink.extract(state)
-            (b, leftovers) = extractResult
-            nextState = if (leftovers.isEmpty) State.Empty(initial, notifyConsumer)
-            else State.Leftovers(initial, leftovers, notifyConsumer)
-            // Inform the producer that we took the batch, extract the sink and emit the data
-          } yield (notifyProducer.succeed(()).as(Chunk.single(b)), nextState)).mapError(Some(_))
-
-        case State.BatchEnd(state, notifyProducer) =>
-          (for {
-            initial        <- sink.initial
-            notifyConsumer <- Promise.make[Nothing, Unit]
-            extractResult  <- sink.extract(state)
-            (b, leftovers) = extractResult
-            nextState = if (leftovers.isEmpty) State.Empty(initial, notifyConsumer)
-            else State.Leftovers(initial, leftovers, notifyConsumer)
-          } yield (notifyProducer.succeed(()).as(Chunk.single(b)), nextState)).mapError(Some(_))
-
-        case e @ State.Error(cause) => ZIO.succeed((ZIO.halt(cause.map(Some(_))), e))
-        case State.End              => ZIO.succeed((ZIO.fail(None), State.End))
+      withStateVar(stateVar, permits) { s =>
```

Comment:
has anything changed here except the addition of the explicit match? If not, let's roll it back so the diff is smaller